### PR TITLE
Add support for nested list types

### PR DESCRIPTION
**Pull request by nunezco2**
This code modifies the mechanism by which BaseAnalysis validates parameters via a specification when these are lists. Using tuples, the type of the container (list, set) can be checked along side the type of the elements.

To test, change the test_meandamage_bridge.py and set one of the damage states "DS_2" to 2.

### CODE REVIEW

File: pyincore/analyses/meandamage/meandamage.py [hunk 1]
```diff
@@ -30,17 +30,17 @@ def get_spec(self):
                     'description': 'result dataset name',
                     'type': str
                 },
+                {
+                    'id': 'damage_interval_keys',
+                    'required': True,
+                    'description': 'Column name of the damage interval must be four and ranged in order',
+                    'type': (list, str)
```
Comment (longshuicy) [inline pyincore/analyses/meandamage/meandamage.py:37]: hmm this representation might be a little bit hard to understand. Could we do `list<str>` instead? You are parsing the tuple anyway

**Review by longshuicy — APPROVED** [review]
If you have time, could you add a test for that?

File: pyincore/baseanalysis.py [hunk 2]
```diff
@@ -143,6 +144,24 @@ def set_output_dataset(self, id, dataset):
             # TODO handle error message
             return False
 
+    def validate_parameter_nested(self, parameter, parameter_spec):
+        is_valid = True
+        err_msg = ''
+
+        if type(parameter_spec['type']) is typing._GenericAlias:
```
Comment (navarroc) [inline pyincore/baseanalysis.py:151]: there is a warning about accessing a protected member _GenericAlias, is there any concern here of using this or is there a better way?

File: pyincore/baseanalysis.py [hunk 3]
```diff
@@ -156,16 +175,16 @@ def validate_parameter(self, parameter_spec, parameter):
...
-            elif not type(parameter) is parameter_spec['type']:
-                is_valid = False
-                err_msg = 'parameter type does not match - spec: ' + str(parameter_spec)
-        elif not isinstance(parameter, type(None)) and not (type(parameter) is parameter_spec['type']):
-            is_valid = False
-            err_msg = 'parameter type does not match - spec: ' + str(parameter_spec)
+            else:
+                is_valid, err_msg = self.validate_parameter_nested(parameter, parameter_spec)
+        else:
+            if not parameter is None:
```
Comment (navarroc) [inline pyincore/baseanalysis.py:186]: pep8 error about using "not"

**Comment by navarroc** [comment]
Since we reviewed this together last week, I think we have consensus to get this into this release once we address a few things:

1. We need to merge in develop to fix the tests
2. There was a pep8 error I noted
3. The question about accessing the protected variable

Once the above are addressed, I think we can merge this. We'll also need to remember to update the documentation for the analyses so we change "list" to "List[type]".

**Review by longshuicy — APPROVED** [review]

**Comment by navarroc** [comment]
Did you try running any of analyses to verify the id change you made? Unfortunately IN-CORE is down today so we can't do that until it's back online.

**Comment by nunezco2** [comment]
> Did you try running any of analyses to verify the id change you made? Unfortunately IN-CORE is down today so we can't do that until it's back online.

Yes, I did try the change, but let's better wait for the services to be up again so others can test it more thoroughly.

**Comment by navarroc** [comment]
> > Did you try running any of analyses to verify the id change you made? Unfortunately IN-CORE is down today so we can't do that until it's back online.
> 
> Yes, I did try the change, but let's better wait for the services to be up again so others can test it more thoroughly.

It's in the incore-docs repository in the MD files for each analysis.

File: pyincore/analyses/epfrestoration/epfrestoration.py [hunk 4]
```diff
@@ -9,6 +9,7 @@
```
Comment (navarroc) [inline pyincore/analyses/epfrestoration/epfrestoration.py:9]: Not related to this PR, but can you add the copyright 2022 line to the header for epfrestoration?

**Review by navarroc — APPROVED** [review]

File: pyincore/analyses/waterfacilityrestoration/waterfacilityrestoration.py [hunk 5]
```diff
@@ -9,6 +9,7 @@
 
 import numpy as np
 
+from typing import List
```
[the remark this answers is not in the document]
Comment (nunezco2) [reply pyincore/analyses/waterfacilityrestoration/waterfacilityrestoration.py:12]: Yes, will do

**Pull request merged**